Add `openWith` function to OCA.Viewer

Raudius opened the pull request:
Adds the function `OCA.Viewer.openWith(handler, filePath, options)`

The function can be used for opening files with the specified handler. This functionality would be needed for cases where a handler may not be suitable as the default option for a mime-type (e.g. collabora and PDFs https://github.com/nextcloud/richdocuments/issues/1894)

skjnldsv commented:
So, you would just register a new action called `Open with Collabora` for example?
Would you share some implementations or a bit more code?

I assume this is an issue about default registration priority?

Raudius commented:
@skjnldsv 
> So, you would just register a new action called Open with Collabora for example?

Yes, there is a separate action in the Files app for PDF files that should open the file in the richdocuments viewer.

> Would you share some implementations or a bit more code?

With the current PR implementation  we can do something like this when clicking on the `Open with Collabora` button, and this will open the file with the Office component (assuming it has been registered):

```js
OCA.Viewer.openWith({
    id: 'richdocuments',
    component: Office,
    theme: 'light',
}, context.fileInfoModel.getFullPath())
```

> I assume this is an issue about default registration priority?

Right now in Collabora we don't register the PDF mime with the viewer, but we want to open PDFs if explicitly chosen through the actions menu.

skjnldsv commented on src/views/Viewer.vue:
```diff
@@ -383,10 +386,22 @@ export default {
+					handler = this.components[mime]
+				}
+
+				if (overrideHandler !== null) {
```
```suggestion
				// Try provided handler if any
				if (overrideHandler !== null) {
```

skjnldsv commented on src/services/Viewer.js:
```diff
@@ -189,6 +199,25 @@ export default class Viewer {
+	 * @param {Function} options.onClose callback when closing the viewer
+	 */
+	openWith(handler, options = {}) {
+		this._state.overrideHandler = handler
```
Make sure to reset this on Close

Raudius commented:
@skjnldsv 
Thanks for the review, after your feedback I changed some things around.

It didnt make much sense to pass the full handler to the openWith function so I changed it so we only need to pass the handler ID:
```js
OCA.Viewer.openWith(
  'richdocuments', 
  { path: context.fileInfoModel.getFullPath() }
)
```

Also previously I immediately cleared the override handler after opening the file. But it makes more sense to clear it (as you suggested) when the viewer is cleared, in the case that a file-list is passed to the viewer.

szaimen commented:
/compile amend /

juliushaertl commented:
@skjnldsv Mind to have another quick look as well?

eneiluj commented:
/compile amend /

eneiluj reviewed (approved):
I took the liberty to add https://github.com/nextcloud/viewer/pull/1273/commits/1ac581fd8acbcf13fc30bba5c99ae9bcff53d5c9
to avoid confusion between the selected handler and the one in the find lambda function.

Works fine with https://github.com/nextcloud/richdocuments/pull/2298 :+1: 
Code looks good.

Raudius commented:
/backport to 24
